js/piwik.js

```diff
@@ -2605,10 +2609,34 @@ if (typeof Piwik !== 'object') {
                 callback();
             }
 
+            /**
+             * Send each bulk tracks to the server
+             * @return {void}
+             */
+            function sendBulkTrackingRequests() {
+
+                // Work on a local copy
+                var tmp = JSON.parse(JSON.stringify(bulkTrackingCollection));
+                bulkTrackingCollection.length = 0;
+
+                // Do a request only if we have to
+                if(tmp.length >= 1) {
+                  var bulk = '{"requests":["?' + tmp.join('","?') + '"]}';
+                  sendRequest(bulk, configTrackerPause, null, true);
+                }
+            }
+
             /*
              * Send request
              */
-            function sendRequest(request, delay, callback) {
+            function sendRequest(request, delay, callback, force) {
+
+
+                if(enableBulkTracking && !force) {
+                  bulkTrackingCollection.push(request);
+                  return false;
```

Comment:
A few lines further below is the `heartBeatSetUp`. By returning here it can never reach the heart beat see https://github.com/piwik/piwik/blob/2.15.0/js/piwik.js#L3188 . This is a bit of a design problem of us here since the `heartBeatSetup` shouldn't be in `sendRequest()`. Can you maybe refactor the code slightly? Like `if (bulk) { request.push() } else { sendrequest normal way }... setupheartbeat...`